docs/specs/pipelines-io.md

```diff
@@ -0,0 +1,196 @@
+   public void Advance(int bytesWritten);
+   public void Commit();
+   public Task FlushAsync();
+}
+```
+
+There are extension methods like `Write(ReadOnlySpan<byte>)` which automate writing to the available memory and advancing the buffer.
+
+```C#
+public static void Write(this WritableBuffer buffer, ReadOnlySpan<byte> source)
+{
+    if (buffer.Memory.IsEmpty)
+    {
+        buffer.Ensure();
+    }
+
+    // Fast path, try copying to the available memory directly
+    if (source.Length <= buffer.Memory.Length)
+    {
+        source.CopyTo(buffer.Memory.Span);
+        buffer.Advance(source.Length);
+        return;
+    }
+
+    var remaining = source.Length;
+    var offset = 0;
+
+    while (remaining > 0)
+    {
+        var writable = Math.Min(remaining, buffer.Memory.Length);
+
+        buffer.Ensure(writable);
+
+        if (writable == 0)
+        {
+            continue;
+        }
+
+        source.Slice(offset, writable).CopyTo(buffer.Memory.Span);
+
+        remaining -= writable;
+        offset += writable;
+
+        buffer.Advance(writable);
+    }
+}
+```
+
+### Reading
+
+The consumer calls `IPipelineReader.ReadAsync()` which returns with a `ReadableBuffer` when the producer says that data is ready. 
+
+```C#
+private Task Consume(IPipelineReader reader)
+{
+    while (true)
+    {
+        ReadResult result = await reader.ReadAsync();
+        ReadableBuffer buffer = result.Buffer;
+        if (buffer.IsEmpty && result.IsCompleted)
+        {
+            break;
+        }
+
+        // Process the buffer
+        Consume(ref buffer);
+        
+        // Advance the reader past the entire buffer
+        reader.Advance(buffer.End, buffer.End);
+    }
+}
+```
+
+A `ReadableBuffer` is a struct that is a view over one or many `BufferSegments`:
+
+```C#
+public struct ReadCursor
+{
+   public int Index { get; }
+   public BufferSegment Segment { get; }
+}
+
+public struct ReadableBuffer : ISequence<Memory<byte>>
+{
+   public ReadCursor Start { get; }
+   public ReadCursor End { get; }
+   public int Length { get; }
+   
+   public byte Peek();
+   public ReadableBuffer Slice(ReadCursor start, ReadCursor end);
+   public PreservedBuffer Preserve();
+}
+
+public struct PreservedBuffer : IDisposable
+{
+   public ReadableBuffer Buffer { get; }
+}
+```
+
+A `ReadCursor` represents a specific coordinate within a `BufferSegment` and a `ReadableBuffer` represents all of the data between 2 `ReadCursors`, Start and End. The `ReadableBuffer` is a point in time snapshot of the data the producer made available. 
+
+After processing the buffer, consumers call `IPipelineReader.Advance(consumed, observed)` indicate how much of the buffer was consumed and how much of the buffer was observed. The distinction is very important as consumed data can be returned to the system for re-use by another component. Marking observed data lets the consumer control over where it should be notified again (to an extent). 
```

Comment:
Did you considered removing "observed" and assuming that everything in the buffer was observed? At least this being the default that could be explicitly changed, e.g. two overloads of Advance, one with used consumed.